x/ibc/handler.go

```diff
@@ -192,8 +192,12 @@ func NewHandler(k keeper.Keeper) sdk.Handler {
 				return nil, sdkerrors.Wrap(err, "receive packet callback failed")
 			}
 
+			if err := k.ChannelKeeper.ReceiveExecuted(ctx, cap, msg.Packet); err != nil {
+				return nil, err
+			}
+
 			// Set packet acknowledgement
-			if err = k.ChannelKeeper.ReceiveExecuted(ctx, cap, msg.Packet, ack); err != nil {
+			if err := k.ChannelKeeper.WriteAcknowledgement(ctx, msg.Packet, ack); err != nil {
```

Comment:
I don't think this should be called in the handler? Acknowledgements should be allowed to written after this handler call. `WriteAcknowledgement` should function in the same way as `SendPacket`, the application modules calls it at some point during its execution

Reply:
If there is a desire not to break the API compatibility with existing `OnRecvPacket`, then we should allow the `ack` to be returned, but only write it if it is not `nil`.  If `nil`, then allow the app to call `WriteAcknowledgement` later.

Not sure how desirable this compatibility is.

Reply:
cc/ @cwgoes